elasticsearch/src/main/java/org/hibernate/search/elasticsearch/util/impl/GsonHttpEntity.java

```diff
@@ -192,81 +235,133 @@ public void fillDigest(MessageDigest digest) throws IOException {
 			digestWriter.insertNewline();
 		}
 		//Now we finally know the content size in bytes:
-		this.contentLength = digestWriter.getContentLenght();
+		hintContentLenght( digestWriter.getContentLenght() );
+	}
+
+	private void hintContentLenght(long contentLenght) {
+		if ( contentlengthWasProvided == false ) {
+			this.contentLength = contentLenght;
+		}
 	}
 
-	private static final class ProgressiveByteBufWriter extends Writer {
+	private static final class ProgressiveCharBufferWriter extends Writer {
 
-		//Initially null: must be set before writing,
-		//as it might change between writes.
+		//Initially null: must be set before writing is starter and each
+		//time it's resumed as it might change between writes during
+		//chunked encoding.
 		ContentEncoder out;
 
 		//This is to keep the buffers which could not be written to socket
 		//because of flow control. Make sure to eventually push them out.
 		//Initialised only if/when needed.
-		List<ByteBuffer> unwrittenLazyBuffers;
+		List<CharBuffer> unwrittenLazyBuffers;
 
 		//Set this to true when we detect clogging, so we can stop trying.
 		//Make sure to reset this when the HTTP Client hints so.
 		//It's never dangerous to re-enable, just not efficient to try writing
 		//unnecessarily.
 		boolean flowControlPushingBack = false;
 
-		//Accumulate the content length while it's being produced so we can hint
-		//other components.
-		long totalWrittenBytes = 0;
+		private final CharsetEncoder utf8Encoder = StandardCharsets.UTF_8.newEncoder();
+		private ByteBuffer availableBuffer = ByteBuffer.allocate( BUFFER_SIZES );
+		private ByteBuffer needsWritingBuffer;
 
 		@Override
 		public void write(char[] cbuf, int off, int len) throws IOException {
 			CharBuffer toWriteChars = CharBuffer.wrap( cbuf, off, len );
-			ByteBuffer toWriteBytes = StandardCharsets.UTF_8.encode( toWriteChars );
-			writeOrBufferWrite( toWriteBytes );
-		}
-
-		public long getTotalWrittenAndReset() {
-			final long total = totalWrittenBytes;
-			//To not count size multiple times on repeated renders:
-			totalWrittenBytes = 0;
-			return total;
+			if ( flowControlPushingBack == false ) {
+				writeOrStore( toWriteChars, false, true );
+			}
+			else {
+				storeForLater( toWriteChars );
+			}
 		}
 
 		public void resumePendingWrites() throws IOException {
 			flowControlPushingBack = false;
-			flushPendingBuffers();
+			if ( availableBuffer == null ) {
+				attemptFlushPendingBuffers( false );
+			}
 		}
 
-		public void insertNewline() throws IOException {
-			writeOrBufferWrite( ByteBuffer.wrap( NEWLINE ) );
+		private void storeForLater(CharBuffer toWriteLater) {
+			if ( unwrittenLazyBuffers == null ) {
+				unwrittenLazyBuffers = new ArrayList<>( 16 );
+			}
+			//The original CharBuffer is not safe as it's backed by a char array
+			//which happens to be the GSON write buffer: it will get mutated.
+			CharBuffer defensiveCopy = CharBuffer.allocate( toWriteLater.remaining() );
```

Comment:
I'm not sure I understand what you'd hope to achieve. I'd love to simplify the code a bit but not at cost of performance, and I'm not sure if you've considered all aspect .. for example that you actually don't know how large the bytebuffers need to be as that depends on the encoding details, and guessing gets you into re-sizing trouble.

I need to wrap this up, let's apply this as-is but then if you want to play with it and show me I'm happy to review some follow-up polishing PR.